packages/dashboard/src/dashboard-actions/redo.spec.ts

```diff
@@ -0,0 +1,182 @@
+    const emptyDashState: DashboardReducerState = {
+      dashboardConfiguration: {
+        widgets: [],
+        viewport: { duration: '5m' },
+      },
+      selectedWidgetIds: [],
+      numTimesCopyGroupHasBeenPasted: 0,
+      copyGroup: [],
+      stretchToFit: true,
+      width: 1000,
+      cellSize: 30,
+      isActionComplete: true,
+    };
+    expect(
+      redo({
+        dashboardAction: moveAction,
+        dashboardState: emptyDashState,
+      })
+    ).toEqual(emptyDashState);
+  });
+
+  it('reverses a single move', () => {
+    const moveAction = onMoveAction({
+      position: { x: MOCK_KPI_WIDGET.x, y: MOCK_KPI_WIDGET.y },
+      prevPosition: { x: 50, y: 50 },
+      widgetIds: ['mock-kpi-widget'],
+      cellSize: 30,
+    });
+    moveAction.payload.prevPosition as Position;
+    const reversedDashboard = move({
+      position: moveAction.payload.prevPosition,
+      previousPosition: moveAction.payload.position,
+      cellSize: moveAction.payload.cellSize,
+      selectedWidgetIds: moveAction.payload.widgetIds,
+      dashboardConfiguration: state.dashboardConfiguration,
+    });
+
+    expect(
+      redo({
+        dashboardAction: reverseMove(moveAction),
+        dashboardState: state,
+      }).dashboardConfiguration
+    ).toEqual(reversedDashboard);
+  });
+});
+
+describe('RESIZE', () => {
+  it('returns original state reversed twice', () => {
+    const resizeAction: ResizeAction = onResizeAction({
+      anchor: 'right',
+      changeInPosition: { x: 10, y: 10 },
+      widgetIds: [MOCK_KPI_WIDGET.id],
+      cellSize: state.cellSize,
+      dashboardConfiguration: state.dashboardConfiguration,
+    });
+
+    expect(
+      redo({
+        dashboardAction: reverseResize(resizeAction),
+        dashboardState: redo({
+          dashboardAction: resizeAction,
+          dashboardState: state,
+        }),
+      })
+    ).toEqual(state);
+  });
+});
+
+describe('PASTE', () => {
+  const pasteState: DashboardReducerState = {
+    dashboardConfiguration: {
+      widgets: [MOCK_KPI_WIDGET],
+      viewport: { duration: '5m' },
+    },
+    selectedWidgetIds: [],
+    numTimesCopyGroupHasBeenPasted: 1,
+    copyGroup: [MOCK_KPI_WIDGET],
+    stretchToFit: true,
+    width: 1000,
+    cellSize: 10,
+    isActionComplete: true,
+  };
+  pasteState.dashboardConfiguration = paste({
+    dashboardConfiguration: pasteState.dashboardConfiguration,
+    copyGroup: state.copyGroup,
+    numTimesCopyGroupHasBeenPasted: state.numTimesCopyGroupHasBeenPasted,
+  });
+
+  it('returns original state when reversed twice', () => {
+    const pasteState: DashboardReducerState = {
+      dashboardConfiguration: {
+        widgets: [MOCK_KPI_WIDGET],
+        viewport: { duration: '5m' },
+      },
+      selectedWidgetIds: [],
+      numTimesCopyGroupHasBeenPasted: 1,
+      copyGroup: [MOCK_KPI_WIDGET],
+      stretchToFit: true,
+      width: 1000,
+      cellSize: 10,
+      isActionComplete: true,
```

Comment:
is `isActionComplete` something needs to be stored as state?

Reply:
no it definitely doesn't have to be stored there, fixing that.